File: 700/715 - Range Module.java
import java.util.HashSet;
import java.util.Set;
import java.util.TreeMap;

class RangeModule {

    private TreeMap<Integer, Integer> map;
    
    public RangeModule() {
        map = new TreeMap<>();
    }
    
    private int[] findInsert(int left, int right) {
        Set<Map.Entry<Integer, Integer>> removeEntries = new HashSet<>();
        int l = left, r = right;
        
        Map.Entry<Integer, Integer> lleft = map.floorEntry(left);
        if (lleft != null && lleft.getValue() >= left) {
            removeEntries.add(lleft);
            l = lleft.getKey();
            r = Math.max(lleft.getValue(), r);
        }
        
        for (Map.Entry<Integer, Integer> entry : map.subMap(left, right + 1).entrySet()) {
            r = Math.max(entry.getValue(), r);
            
            removeEntries.add(entry);
        }
        
        map.entrySet().removeAll(removeEntries);

        return new int[] {l, r};
    }
    
    public void addRange(int left, int right) {
        int[] range = findInsert(left, right);
        map.put(range[0], range[1]);
    }
    
    public boolean queryRange(int left, int right) {
        Map.Entry<Integer, Integer> l = map.floorEntry(left);
        return l != null && l.getValue() >= right;
    }
    
    public void removeRange(int left, int right) {
        int[] range = findInsert(left, right);

        if (left > range[0]) {
            map.put(range[0], left);
        }
        
        if (right < range[1]) {
            map.put(right, range[1]);
        }
    }
}

/**
 * Your RangeModule object will be instantiated and called as such:
 * RangeModule obj = new RangeModule();
 * obj.addRange(left,right);
 * boolean param_2 = obj.queryRange(left,right);
 * obj.removeRange(left,right);
 */

File: csharp/715.range-module.cs
/*
 * @lc app=leetcode id=715 lang=csharp
 *
 * [715] Range Module
 */

using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class RangeModule {

    private IList<(int left, int right)> Ranges;

    public RangeModule() {
        Ranges = new List<(int left, int right)>();
    }
    
    public void AddRange(int left, int right) {
        var tmp = new List<(int left, int right)>();
        int n = Ranges.Count, i = 0;
        while (i < n) {
            var range = Ranges[i];
            if (range.right >= left) {
                break;
            }
            tmp.Add(range);
            ++i;
        }
        while (i < n) {
            var range = Ranges[i];
            if (range.left > right) {
                break;
            }
            left = Math.Min(left, range.left);
            right = Math.Max(right, range.right);
            ++i;
        }
        tmp.Add((left: left, right: right));
        for (int j = i; j < n; ++j) {
            tmp.Add(Ranges[j]);
        }
        Ranges = tmp;
    }
    
    public bool QueryRange(int left, int right) {
        return Ranges.TakeWhile(range => range.left <= right).Any(range => range.left <= left && range.right >= right);
    }
    
    public void RemoveRange(int left, int right) {
        var tmp = new List<(int left, int right)>();
        int n = Ranges.Count, i = 0;
        while (i < n) {
            var range = Ranges[i];
            if (range.right > left) {
                break;
            }
            tmp.Add(range);
            ++i;
        }
        while (i < n) {
            var range = Ranges[i];
            if (range.left >= right) {
                break;
            }
            if (range.left < left) {
                tmp.Add((left: range.left, right: left));
            }
            if (range.right > right) {
                tmp.Add((left: right, range.right));
            }
            ++i;
        }
        for (int j = i; j < n; ++j) {
            tmp.Add(Ranges[j]);
        }
        Ranges = tmp;
    }
}

/**
 * Your RangeModule object will be instantiated and called as such:
 * RangeModule obj = new RangeModule();
 * obj.AddRange(left,right);
 * bool param_2 = obj.QueryRange(left,right);
 * obj.RemoveRange(left,right);
 */
// @lc code=end

